<table>
    <tr>
        <td><img src="https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/logo_ECI.png" width="300"/></td>
        <td>&nbsp;</td>
        <td>
            <h1 style="font-size:200%;color:blue;text-align:center">    <FONT COLOR="blue">  Sesión 2 </p> Fundamentación </p> Análisis Exploratorio de Datos </p>    </FONT>         </h1></td>         
        <td>
            <tp><p style="font-size:99%;text-align:center">Seminario IA-Salud </p></tp>
            <tp><p style="font-size:115%;text-align:center">Curso 2026-1</p></tp>
            <tp><p style="font-size:115%;text-align:center">Prof. Fabián Sánchez</p></tp>
        </td>
    </tr>
</table>

In [1]:
%%html
<center><marquee style='width: 60%; color: red'><b>✌ ¡Hola a todos y a todas! Un saludo... Bienvenidos a la introducción a Python-Colab✌ </b></marquee><center>

# <FONT SIZE=5 COLOR="green"> 1. Python </FONT>

Python es un lenguaje de programación de propósito general, usado ampliamente para

*  Desarrollo de aplicaciones
*  Desarrollo web
*  Web Scraping
*  **Ciencia de Datos**
*  **Visualización**
*  Aplicaciones a los negocios
*  **Machine Learning**
*  **Deep Learning**

Python cuenta con librerías especializadas dependiendo del tipo de tarea. Algunas de ellas son:

<center><img src="https://github.com/Fabian830348/cursos/blob/master/librerias.png?raw=true" alt="centered image" width="550" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: https://www.actumlogos.com/python-para-ia.html  </FONT> <figcaption></center>

1.   **Pandas:** manipulación de datos-estadísticas-data.frames
2.   **Matplotlib**: visualización-gráficas
3.   **numpy**: operaciones matemáticas
4.   **spicy**: realizar cálculo simbólico
5.   **scikit-learn**: ejecutar algoritmos de machine learning
6.   **NTKL**: Procesamiento de Lenguaje Natural.
7.   **seaborn**: creación de gráficos estadísticos
8.   **Datetime**: para manipulación de fechas



---



<center><FONT SIZE=5 COLOR="purple"> EXPLORACIÓN DE DATOS EN PYTHON</FONT></center>

<center><img src="https://miro.medium.com/v2/resize:fit:720/format:webp/0*fto5mNElXLZ8TCZr.png" alt="centered image" width="500" height="250"></center>

+ Librería de Python especializada manipulación y análisis de estructuras de datos
+ Permite leer datos en diferentes formatos: csv, excel, bases de datos SQL.
+ Realiza operaciones rápidas y eficientes sobre conjuntos de datos
+ Permite trabajar series de tiempo
+ *import pandas as pd*
+ [Guía Pandas](https://aprendeconalf.es/docencia/python/manual/pandas/)



<center><FONT SIZE=5 COLOR="blue"> DATAFRAMES </FONT></center>

Los tipos de datos que vamos a trabajar se denominan *dataframes*

<center><img src="https://github.com/Fabian830348/cursos/blob/master/dataframe.png?raw=true" alt="centered image" width="600" height="300"></center>

Observe que un *dataframe* puede tener diferentes tipos de variables

<center><img src="https://github.com/Fabian830348/cursos/blob/master/tiposvariables.png?raw=true" alt="centered image" width="600" height="400"></center>





---



<center><FONT SIZE=5 COLOR="PURPLE"> PROCESO DE EXPLORACIÓN DE LOS DATOS </FONT></center>

<br>

<center><img src="https://github.com/Fabian830348/cursos/blob/master/imagen_eda.png?raw=true" alt="centered image" width="550" height="350"></center><center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia </FONT> <figcaption></center>

En el gráfico anterior se muestran algunas etapas fundamentales:

1.	**Preguntas para responder**. Es importante tener claro para qué queremos hacer el procesamiento y análisis de datos. Es decir, qué buscamos responder.

2.	**Generalidades del conjunto de datos**.  Acá buscamos reconocer la dimensión, datos faltantes, y en general, hacer un resumen del conjunto de datos.

3.	**Tipos de Variables**. Clasificar las variables en cualitativas y cuantitativas y examinar como se presenta esta información. Por ejemplo, una variable puede estar codificada y verse como numérica, pero ser categórica.

4.	**Estadística Descriptiva**. Se busca resumir la información. Utilizar medidas de tendencia, de dispersión, de forma, de posición, etc., para entender mejor los datos. Por ejemplo, se identifican datos atípicos usando boxplot.

5.	**Visualización**.  Es una etapa fundamental en el análisis de datos. Visualizar la información y poder describirla gráficamente. En esta etapa se usan: diagramas de barras, pie, boxplot, histogramas, líneas de densidad, diagramas de dispersión, entre otros.

6.	**Relación entre variable**. Se exploran datos bivariados. Se examinan relaciones entre variables cualitativas (tablas de contingencia) y relaciones entre variables cuantitativas (diagramas de dispersión) se establecen posibles relaciones lineales.

7.	**Reporte de resultados**. Una vez se haga el proceso de exploración y análisis en Python, R-Studio u otro, se deben reportar los resultados y conclusiones de lo que se encontró en el conjunto de datos.

<center><FONT SIZE=5 COLOR="blue"> ¿Cómo importamos una librería para trabajar?</FONT></center>

In [5]:
# librería para manejo de dataframes
import pandas as pd
# librería para hacer ciertos cálculos y funciones
import numpy  as np

# <FONT SIZE=5 COLOR="green"> 2. Carga de Datos en Python </FONT>

Para cargar datos en Google-Colab podemos usar varias alternativas, por ejemplo

- Desde el drive

- Repositorio externo como Github u otro.

Se pueden cargar datos en diferentes formatos, por ejemplo:

- $read\_csv$ : archivos separados por comas

- $read\_excel$: clásicas hojas de cálculo

- $read\_json$ : javascript object notation. (formato java)

In [11]:
hospital = pd.read_excel('/content/BBDD_Hospitalizacion.xlsx')

In [13]:
hospital

,EDAD,DIABETES,HOSPITALIZACIÓN ULTIMO MES,PSA,BIOPSIAS PREVIAS,VOLUMEN PROSTATICO,ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS,NUMERO DE MUESTRAS TOMADAS,CUP,ENF. CRONICA PULMONAR OBSTRUCTIVA,BIOPSIA,NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA,FIEBRE,ITU,TIPO DE CULTIVO,AGENTE AISLADO,PATRON DE RESISTENCIA,HOSPITALIZACION,DIAS HOSPITALIZACION MQ,DIAS HOSPITALIZACIÓN UPC
0,53.0,NO,NO,4.00,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,2.0,0.0
1,56.0,NO,NO,7.70,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
2,57.0,NO,NO,29.00,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,"AMPI R, CIPRO R, GENTA R, SULFA M R",SI,4.0,3.0
3,56.0,NO,NO,7.00,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
4,55.0,NO,NO,29.00,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,NO,SI,4.0,3.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
565,69.0,NO,NO,4.43,NaN,NaN,FLUOROQUINOLONA_AMINOGLICOSIDO,20.0,NaN,NO,NEG,NO,NO,NO,NO,NO,NO,NO,0.0,0.0
566,67.0,NO,NO,8.00,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NaN,ADENOCARCINOMA GLEASON 7,NO,NO,NO,NO,NaN,NO,NO,0.0,0.0
567,59.0,NO,NO,9.00,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NaN,NEG,NO,NO,NO,NO,NO,NO,NO,0.0,0.0
568,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [17]:
# 1. revisar los primeros datos (cabeza de los datos)
hospital.head()

,EDAD,DIABETES,HOSPITALIZACIÓN ULTIMO MES,PSA,BIOPSIAS PREVIAS,VOLUMEN PROSTATICO,ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS,NUMERO DE MUESTRAS TOMADAS,CUP,ENF. CRONICA PULMONAR OBSTRUCTIVA,BIOPSIA,NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA,FIEBRE,ITU,TIPO DE CULTIVO,AGENTE AISLADO,PATRON DE RESISTENCIA,HOSPITALIZACION,DIAS HOSPITALIZACION MQ,DIAS HOSPITALIZACIÓN UPC
0,53.0,NO,NO,4.0,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,2.0,0.0
1,56.0,NO,NO,7.7,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
2,57.0,NO,NO,29.0,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,"AMPI R, CIPRO R, GENTA R, SULFA M R",SI,4.0,3.0
3,56.0,NO,NO,7.0,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
4,55.0,NO,NO,29.0,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,NO,SI,4.0,3.0


In [18]:
# 2, visualizar los ultimos datos
hospital.tail()

,EDAD,DIABETES,HOSPITALIZACIÓN ULTIMO MES,PSA,BIOPSIAS PREVIAS,VOLUMEN PROSTATICO,ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS,NUMERO DE MUESTRAS TOMADAS,CUP,ENF. CRONICA PULMONAR OBSTRUCTIVA,BIOPSIA,NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA,FIEBRE,ITU,TIPO DE CULTIVO,AGENTE AISLADO,PATRON DE RESISTENCIA,HOSPITALIZACION,DIAS HOSPITALIZACION MQ,DIAS HOSPITALIZACIÓN UPC
565,69.0,NO,NO,4.43,NaN,NaN,FLUOROQUINOLONA_AMINOGLICOSIDO,20.0,NaN,NO,NEG,NO,NO,NO,NO,NO,NO,NO,0.0,0.0
566,67.0,NO,NO,8.00,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NaN,ADENOCARCINOMA GLEASON 7,NO,NO,NO,NO,NaN,NO,NO,0.0,0.0
567,59.0,NO,NO,9.00,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NaN,NEG,NO,NO,NO,NO,NO,NO,NO,0.0,0.0
568,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
569,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [23]:
# 3. tamaño de los datos
hospital.shape

(570, 20)

In [27]:
# 4. nombre de variables o columnas
hospital.columns

Index(['EDAD', 'DIABETES', 'HOSPITALIZACIÓN ULTIMO MES', 'PSA',
       'BIOPSIAS PREVIAS', 'VOLUMEN PROSTATICO',
       'ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS', 'NUMERO DE MUESTRAS TOMADAS',
       'CUP', 'ENF. CRONICA PULMONAR OBSTRUCTIVA', 'BIOPSIA',
       'NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA',
       'FIEBRE', 'ITU', 'TIPO DE CULTIVO', 'AGENTE AISLADO',
       'PATRON DE RESISTENCIA', 'HOSPITALIZACION', 'DIAS HOSPITALIZACION MQ',
       'DIAS HOSPITALIZACIÓN UPC'],
      dtype='object')

In [29]:
# 5. info
hospital.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 20 columns):
 #   Column                                                                     Non-Null Count  Dtype  
---  ------                                                                     --------------  -----  
 0   EDAD                                                                       568 non-null    float64
 1   DIABETES                                                                   568 non-null    object 
 2   HOSPITALIZACIÓN ULTIMO MES                                                 568 non-null    object 
 3   PSA                                                                        564 non-null    float64
 4   BIOPSIAS PREVIAS                                                           566 non-null    object 
 5   VOLUMEN PROSTATICO                                                         567 non-null    object 
 6   ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS                   

In [30]:
# 6. estadisticos descriptivos generales
# only en numerical variables
# std desviacion estandar: puntos de desciacion a partir de la mean:media, para arriba y para abajo.
# 25%, 50% y 75%, por ejemplo si
# Q1 = 59 es que el 25% de los pacientes estan entre 39 y 59
# Q2=50%, por ejemplo Q2= 64: el 50% de la poblacion esta entre 39 y 64 años
# Q3 = 75% por ejemplo edad 69: el 75% de los pacientes estan por debajo de 69 años


hospital.describe()

,EDAD,PSA,NUMERO DE MUESTRAS TOMADAS,DIAS HOSPITALIZACION MQ,DIAS HOSPITALIZACIÓN UPC
count,568.000000,564.000000,568.000000,568.000000,568.000000
mean,63.882042,13.271401,12.996479,0.165493,0.014085
std,9.281517,15.668966,3.921180,0.958066,0.196474
min,39.000000,0.600000,4.000000,0.000000,0.000000
25%,59.000000,5.600000,12.000000,0.000000,0.000000
50%,64.000000,8.000000,12.000000,0.000000,0.000000
75%,69.000000,15.000000,12.000000,0.000000,0.000000
max,151.000000,100.000000,24.000000,12.000000,3.000000


In [33]:
# explorar variables mas a fondo:
hospital['DIABETES'].unique()

array(['NO', 'SI', nan], dtype=object)

In [34]:
# variable antibiotico
hospital["ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS"].unique()

array(['FLUOROQUINOLONA_AMINOGLICOSIDO', 'CEFALOSPORINA_AMINOGLUCOCIDO',
       'OROQUINOLONAS', 'FLUOROQUINOLONA_AMINOGLICÓSIDO', 'OTROS', nan],
      dtype=object)

In [35]:
hospital['ENF. CRONICA PULMONAR OBSTRUCTIVA'].unique()

array(['NO', 'SI, EPOC', 'SI, ASMA', 'SI', nan], dtype=object)

In [38]:
# cuantas personas tuvieron fiebre
hospital['FIEBRE'].value_counts()

,count
FIEBRE,
NO,531
SI,37


In [39]:
hospital['NUMERO DE MUESTRAS TOMADAS'].value_counts()

,count
NUMERO DE MUESTRAS TOMADAS,
12.0,432
24.0,42
18.0,36
6.0,26
10.0,10
4.0,6
20.0,4
16.0,4
8.0,2


In [40]:
hospital['ITU'].value_counts()

,count
ITU,
NO,562
SI,6


In [42]:
hospital['HOSPITALIZACION'].value_counts()

,count
HOSPITALIZACION,
NO,541
SI,24


In [43]:
hospital['AGENTE AISLADO'].value_counts()

,count
AGENTE AISLADO,
NO,543
E.COLI,6
PSEUDOMONAS AERUGINOSA,2


In [45]:
#Cruzar variables hospitalizacioncon fiebre: pd.crosstab(variable 1, variable 2)
pd.crosstab(hospital['HOSPITALIZACION'], hospital['FIEBRE'])

FIEBRE,NO,SI
HOSPITALIZACION,,
NO,525,16
SI,3,21


In [46]:
pd.crosstab(hospital['HOSPITALIZACION'], hospital['ITU'])

ITU,NO,SI
HOSPITALIZACION,,
NO,541,0
SI,18,6


In [47]:
hospital.head()

,EDAD,DIABETES,HOSPITALIZACIÓN ULTIMO MES,PSA,BIOPSIAS PREVIAS,VOLUMEN PROSTATICO,ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS,NUMERO DE MUESTRAS TOMADAS,CUP,ENF. CRONICA PULMONAR OBSTRUCTIVA,BIOPSIA,NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA,FIEBRE,ITU,TIPO DE CULTIVO,AGENTE AISLADO,PATRON DE RESISTENCIA,HOSPITALIZACION,DIAS HOSPITALIZACION MQ,DIAS HOSPITALIZACIÓN UPC
0,53.0,NO,NO,4.0,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,2.0,0.0
1,56.0,NO,NO,7.7,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
2,57.0,NO,NO,29.0,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,"AMPI R, CIPRO R, GENTA R, SULFA M R",SI,4.0,3.0
3,56.0,NO,NO,7.0,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
4,55.0,NO,NO,29.0,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,NO,SI,4.0,3.0


In [51]:
# PARA GUARDAR LA TABLA O EXPORTARLA SE PONE .to_csv('nombre de la tabla') y luego se descarga de la barra lateral.
#pd.crosstab(hospital['NUMERO DE MUESTRAS TOMADAS'] , hospital['HOSPITALIZACION']).to_csv('TABLA RESUMEN')

In [ ]:
#TAREA: fabian.sanchez@urosario.edu.co
# TOMAR OTRO CONJUNTO DE DATOS, HACER UN ANALISIS EXPLORATORIO DE DATOS.
# REPLICAR TODOS LOS QUERYS
# POR EJEMPLO CON LA TABLA DE HEPATITIS.
# .head, .tail, .describe,



In [12]:
datos = pd.read_csv('/content/heart.csv')

## <FONT SIZE=4 COLOR="magenta"> **2.1 Desde el drive** </FONT>

Con la siguiente instrucción conectamos google-colab con Drive para poder cargar archivos desde este último.

```python
from google.colab import drive
drive.mount('/content/drive')
```



In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Ya hemos importado la librería ***pandas*** previamente.

**Ejemplo 1**

Vamos a importar el conjunto de datos denominado $BBDD\_Hospitalización.xlsx$.

$$pd.read\_excel(''ruta'')$$

In [7]:
hospital = pd.read_excel("/content/drive/MyDrive/7. Seminario de IA en Salud (2026)/BBDD_Hospitalizacion.xlsx")

In [8]:
hospital

,EDAD,DIABETES,HOSPITALIZACIÓN ULTIMO MES,PSA,BIOPSIAS PREVIAS,VOLUMEN PROSTATICO,ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS,NUMERO DE MUESTRAS TOMADAS,CUP,ENF. CRONICA PULMONAR OBSTRUCTIVA,BIOPSIA,NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA,FIEBRE,ITU,TIPO DE CULTIVO,AGENTE AISLADO,PATRON DE RESISTENCIA,HOSPITALIZACION,DIAS HOSPITALIZACION MQ,DIAS HOSPITALIZACIÓN UPC
0,53.0,NO,NO,4.00,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,2.0,0.0
1,56.0,NO,NO,7.70,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
2,57.0,NO,NO,29.00,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,"AMPI R, CIPRO R, GENTA R, SULFA M R",SI,4.0,3.0
3,56.0,NO,NO,7.00,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NO,NEG,1,SI,NO,NO,NO,NO,SI,5.0,0.0
4,55.0,NO,NO,29.00,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,24.0,NO,NO,NEG,1,SI,NO,HEMOCULTIVO,E.COLI,NO,SI,4.0,3.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
565,69.0,NO,NO,4.43,NaN,NaN,FLUOROQUINOLONA_AMINOGLICOSIDO,20.0,NaN,NO,NEG,NO,NO,NO,NO,NO,NO,NO,0.0,0.0
566,67.0,NO,NO,8.00,NO,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NaN,ADENOCARCINOMA GLEASON 7,NO,NO,NO,NO,NaN,NO,NO,0.0,0.0
567,59.0,NO,NO,9.00,SI,SI,FLUOROQUINOLONA_AMINOGLICOSIDO,12.0,NO,NaN,NEG,NO,NO,NO,NO,NO,NO,NO,0.0,0.0
568,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


**Ejemplo 2**

Vamos a importar el conjunto de datos denominado $diabetes.csv$.
$$pd.read\_csv(''ruta'')$$

In [ ]:
diabetes = pd.read_csv("/content/drive/MyDrive/7. Seminario de IA en Salud (2026)/diabetes.csv")
diabetes

**Nota importante:**
Cuando cargamos datos desde el *DRIVE* las rutas de acceso de los conjuntos de datos son los correspondientes al propietario del notebook.

En ese orden de ideas, si vamos a ejecutar las líneas de código en otra cuenta debemos cambiar las rutas de acceso o de lo contrario, producirá un error.

Para sortear esta situación podemos cargar los datos desde un reporsitorio externo, por ejemplo, desde el ***GitHub*** o desde internet directamente.

## <FONT SIZE=4 COLOR="magenta"> **2.2 Repositorio Externo Github** </FONT>

- Ir al Github (en mi caso FABIAN830348)
https://github.com/Fabian830348

- Copiar la ruta de los datos y se pega acá.

In [ ]:
# datos apartamentos
aptos = pd.read_csv("https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/apartamentos.csv")

In [ ]:
aptos

In [ ]:
# datos cancer
ejercicio = pd.read_csv("https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/exercise.csv" )

In [ ]:
ejercicio

**Nota:** Para reconocer instrucciones de Python vamos a trabajar en este *Notebook* con el conjunto de datos *hospital*

In [ ]:
hospital

# <FONT SIZE=5 COLOR="green"> 3. Exploración de los datos </FONT>

Cuando vamos a explorar los datos. Lo primero que debemos entender es el contexto de los datos.

El conjunto de datos corresponde a: Hospitalización de pacientes por complicaciones infecciosas después de Biopsia Prostática.

1. Antecedentes del paciente.

  - **EDAD**: Edad del paciente

2. Morbilidad asociada al paciente.

  - **DIABETES** Pacience con o sin diabetes
  - **HOSPITALIZACIÓN ÚLTIMO MES**: Indica si el paciente fue hospitalizado el último mes
  - **CUP**: Uso de catéter urinario al momento de la biopsia.
  - **ENF. CRÓNICA PULMONAR OBSTRUCTIVA**: Indica si el paciente tiene una enfermedad crónica pulmonar obstructiva.
  - **VOLUMEN PROSTÁTICO**: Indica si el volumen prostático es mayor a $40 \ cm^3$

3. Antecedentes relacionados con la toma de la biopsia.

  - **PSA**: Concentración de PSA en la sangre (antígeno prostático específico)
  - **BIOPSIAS PREVIAS**: Indica si el paciente a tenido biopsias previas.
  - **ANTIBIOTICO USADO EN LA PROFILAXIS**: Indica el tipo de antibiótico
  - **NÚMERO DE MUESTRAS TOMADAS**: Número de muestras tomadas en la biopsia.
  - **BIOPSIA**: Resultado de la Biopsia

4. Complicaciones infecciosas.

  - **NÚMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA**.
  - **FIEBRE**: Indica si el paciente presenta fiebre.
  - **ITU**: Indica si el paciente presenta infección de tracto urinario.
  - **TIPO DE CULTIVO**: Tipo de cultivo encontrado.
  - **AGENTE AISLADO**: Tipo de agente aislado.
  - **PATRÓN DE RESISTENCIA**: Indica si presenta algún patrón de resistencia.
  - **HOSPITALIZACIÓN**: Indica si el paciente fue hospitalizado luego del procedimiento.
  - **DIAS DE HOSPITALIZACIÓN MQ**: Indica los días de hospitalización médico quirúrgico
  - **DIAS DE HOSPÍTALIZACIÓN UPC**: Indica los días de hospitalización en estado crítico


## <FONT SIZE=5 COLOR="green"> 3.1 Instrucciones de exploración general </FONT>

En esta sección revisaremos:

- Los primeros datos
- Los últimos datos
- El tamaño de los datos
- El nombre de las columnas
- Información del conjunto de datos
. Datos faltantes
- Estadísticas generales

1. Usamos la instrucción **head()** para explorar la cabeza de los datos

- Podemos indicar cuántos datos queremos visualizar simplemente con $head(n)$, sonde $n$ indica el número de registros que queremos ver.

In [ ]:
# para visualizar los primeros datos
hospital.head()

2. Usamos la instrucción **tail()** para explorar los últimos datos

In [ ]:
hospital

In [ ]:
# para visualizar los últimos datos
hospital.tail(4)

3. Con la instrucción **shape** miramos el tamaño de nuestros datos

In [ ]:
# para revisar el tamaño de los datos
hospital.shape

es decir, el conjunto de datos tiene 570 registros o filas y 20 columnas.

4. Para explorar nombre de las variables o columnas **columns**

In [ ]:
# nombres de las columnas
hospital.columns

Podemos ver que tenemos columnas como la edad, diabetes, hospitalización, PSA, etc.

5. **info()** es un resumen de las variables y su clasificación. Además, podemos identificar datos faltantes (NA) en cada una de ellas.

In [ ]:
# resumen de los datos
hospital.info()

Podemos observar:

- Conjunto de datos tiene 570 registros.
- Existen variables que no tienen todos los registros. Es decir, tienen datos faltantes
- Tipo de agente aislado es la que más datos faltantes tiene.


6. Para revisar las estadística generales de las variables **describe()**

- Cuantitativas : ***datos.describe()***

- Cualitativas: ***datos.describe(include='object')***

In [ ]:
# resumen estadístico variables cuantitativas
hospital.describe()

In [ ]:
hospital.describe(include = "object")

In [ ]:
# exportar el conjunto de datos hospital a csv
hospital.to_csv("hospital.csv")

In [ ]:
aptos.describe()

**¿Qué puede decir de la tabla anterior?**

In [ ]:
hospital.columns

In [ ]:
# acceder a una variable
hospital["TIPO DE CULTIVO"]

In [ ]:
hospital.HOSPITALIZACION

Podemos ver las categorías que tiene una variable con:
$$datos["variable"].unique()$$

In [ ]:
# las categorias que tiene la variable
hospital["ENF. CRONICA PULMONAR OBSTRUCTIVA"].unique()

In [ ]:
hospital["ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS"].unique()

7. Para acceder a una o varias variable del data.frame podemos usar la instrucción

$$datos[``\text{nombre de la variable}"]$$

$$datos.variable$$

veamos algunos ejemplos


In [ ]:
hospital.columns

In [ ]:
hospital["ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS"]

In [ ]:
# seleccionar la variable precio
hospital["EDAD"]

Si queremos más variables podemos usar:

In [ ]:
hospital[["EDAD", "DIABETES"]]

## <FONT SIZE=5 COLOR="green"> 3.2 Tabla de frecuencias </FONT>

Para hacer el conteo de las categorías o clases de una variables podemos usar la instrucción:

$$datos[``variable"].value\_counts()$$

Ejemplo: Para hacer una tabla de frecuencias de la variable *HOSPITALIZACIÓN*

In [ ]:
# tabla de frecuencia
hospital["HOSPITALIZACION"].value_counts()

In [ ]:
# Ejercicio: hacer la tabla de frecuencia de la variable ENF. CRONICA PULMONAR OBSTRUCTIVA
hospital["ENF. CRONICA PULMONAR OBSTRUCTIVA"].value_counts()

## <FONT SIZE=5 COLOR="green"> 3.3 Tabla de frecuencias cruzadas </FONT>

Para hacer una tabla de frecuencias cruzadas entre dos o más variables usamos la instrucción:

$$pd.crosstab(datos[``variable1"], datos[``variable1"])$$

In [ ]:
# hacer la tabla cruzada de HOSPITALIZACIÓN Y DIABETES
pd.crosstab(hospital["ENF. CRONICA PULMONAR OBSTRUCTIVA"] , hospital["FIEBRE"])

También se pueden relacionar dos o más variables para la *crosstable*

In [ ]:
pd.crosstab([hospital["HOSPITALIZACION"], hospital["FIEBRE"]]  , hospital["DIABETES"])

# <FONT SIZE=5 COLOR="green"> 4. Manejo de DataFrames </FONT>

En esta sección revisaremos algunas operaciones y acciones que son importantes de conocer en un *dataframe*:

- Ordenar por una variable

- Renombrar una variable

- Eliminar una variable

- Agregar una variable

- Agrupar por categorias de una variable y aplicar funciones.

Para repasar

https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

## <FONT SIZE=4 COLOR="magenta"> 4.1 Ordenar según una variable </FONT>

Para ordenar por una variable usamos la instrucción

$$datos.sort\_values(variable, ascending = False) $$

se usa:

- *ascending=True* :  ordena de menor a mayor.
- *ascending=False* : ordena de mayor a menor.


In [ ]:
hospital.head()

In [ ]:
# ordenamos por la variable Edad de menor a mayor
hospital.sort_values("EDAD").head()

## <FONT SIZE=4 COLOR="magenta"> 4.2 Renombrar una variable </FONT>

Podemos renombrar una columna de la siguiente forma usando la instrucción ***rename***

$$ datos.rename(columns=\{key : value, key : value\}, inplace=True) $$

- Para cambiar el nombre de una variable debemos usar un diccionario donde:
    
  - ***key*** son los nombres actuales de las variables.
  - ***value*** : son los nombres nuevos de las variables.

- La instrucción **inplace** se usa para determinar si cambiamos el *dataframe* original.

In [ ]:
hospital.columns

In [ ]:
hospital.rename(columns = {"ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS" : "Anti" } , inplace = True )
# inplace = True ; para que sobreescriba y tenga efecto en adelante.

In [ ]:
hospital.head()

In [ ]:
# vamos a cambiar el nombre largo de la variable NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA por algo más corto
hospital.rename(columns = { "NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA" : "INF_POST_BIOPSIA"  } ).head(5)

In [ ]:
# para que aplique en adelante
hospital.rename(columns = { "NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA" : "INF_POST_BIOPSIA"  } , inplace=True)

In [ ]:
hospital.head()

In [ ]:
# eliminar una variable diabetes
hospital.drop("DIABETES", axis = 1 , inplace = True)

In [ ]:
hospital.head()

## <FONT SIZE=4 COLOR="magenta"> 4.3 Eliminar filas y columnas</FONT>

Para eliminar una columna o una fila usamos la instrucción ***.drop***

$$datos.drop(``nombre", axis = 1, inplace = T \, or \, F)$$

- *axis = 1* : es por columna
- *axis = 0* : es por  fila
- *inplace = T* : sobreescribe en el *dataframe* original.
- *inplace = F* : no sobreescribe. Mantiene el original

In [ ]:
# eliminar la variable DIAS HOSPITALIZACION MQ
hospital.drop("DIAS HOSPITALIZACION MQ", axis = 1 , inplace = True)

In [ ]:
hospital.head(5)

In [ ]:
# Ejercicio: Eliminar las variables : DIAS HOSPITALIZACIÓN UPC ,  ANTIBIOTICO UTILIZADO EN LA PROFILAXIS


## <FONT SIZE=4 COLOR="magenta"> 4.4 Agrupaciones: ***groupby()*** </FONT>

- Se usa para agrupar una variable por sus categorías y luego aplicar una función a estos grupos con respecto a otra variable

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/groupby1.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>


Podemos hacer operaciones con las agrupaciones, así:

$$datos.groupby(``Variable")[``\text{segunda variable}"].funcion()$$

In [ ]:
# aplicación de group_by
hospital.groupby("HOSPITALIZACION")

**¿Qué podemos hacer con esto?**

In [ ]:
hospital.groupby("HOSPITALIZACION").EDAD.describe()

# <FONT SIZE=5 COLOR="green"> 5. Algunas funciones importantes </FONT>



1. Media o promedio: **.mean()**

- Es el "equilibrio" de los datos.
- Se usa para variables cuantitativas.
- Siempre está entre el mínimo y máximo de los datos.
- Es sensible a datos atípicos. Por está razón se debe tener cuidado con su interpretación.

In [ ]:
hospital.EDAD.mean()

2. Mediana: **.median()**

- Es el valor que divide los datos en dos partes iguales (50%-50%)
- Aplica para variables cuantitativas
- Puede o no estar en el conjunto de datos. Depende si se tiene un número par o impar de elementos.
- No se ve afectada por la existencia de datos atípicos. Por está razón algunas veces es usada para imputar valores atípicos.

In [ ]:
hospital.EDAD.median()

3. Moda: **.mode()**

- Es el valor que más se repite. Es decir el que tiene mayor frecuencia
- Aplica para variables cualitativas y cuantitativas
- UN conjunto puede no tener o tener múltiples modas.

In [ ]:
hospital.EDAD.mode()

4. Máximo y mínimo del conjunto: **.max** y **.min**

In [ ]:
hospital.EDAD.max()

In [ ]:
hospital.EDAD.min()

# <FONT SIZE=5 COLOR="green"> 6. Eliminar los NA's </FONT>

Para eliminar los NA´s podemos usar la instrucción

$$datos.dropna(inplace=True)$$

Toca tener especial cuidado en eliminar NA porque se puede reducir el conjunto de datos considerablemente.

In [ ]:
hospital.dropna(inplace=False)

In [ ]:
hospital.info()

In [ ]:
hospital.head(10)

Observe que cuando se elimina una fila el índice también se elimina. En el ejemplo anterior no aparece el índice 9. Para arreglar esta situación usamo la instrucción

$$hospital.reset\_index$$

In [ ]:
hospital.reset_index(inplace=True)

In [ ]:
hospital.info()

# <FONT SIZE=5 COLOR="green"> 7. Algunas gráficas importantes </FONT>

En esta sección revisaremos algunas gráficas clásicas.

Para esto podemos usar diferentes librerías: *mathplotlib* y *plotly*. También se pueden hacer gráficos elaborados en *seaborn* Más documentación la pueden encontrar en los siguientes links.

- https://matplotlib.org/

- https://seaborn.pydata.org/

- https://plotly.com/python/plotly-express/

**Nota:** nos centraremos más en la parte gráfica y el código que en el análisis de las gráficas.

<center><FONT SIZE=5 COLOR="Magenta"> Mathplotlib - Seaborn - Plotly.Express </FONT></center>

Existen múltiples librerías para hacer gráficas en Python. En esta parte mostraremos algunos ejemplos, sin embargo, invitamos al lector a explorar con más detalle estas librerías.

- **Mathplotlib**: es una biblioteca de Python utilizada para trazar gráficos con la ayuda de otras bibliotecas como Numpy y Pandas. Es una poderosa herramienta para visualizar datos en Python. Se utiliza para crear inferencias estáticas y trazar gráficos 2D de matrices. Fue presentado por primera vez por John D. Hunter en 2002.


- **Seaborn**: también es una biblioteca de Python que se utiliza para trazar gráficos con la ayuda de Matplotlib, Pandas y Numpy. Está construido sobre el techo de Matplotlib y se considera como un superconjunto de la biblioteca Matplotlib. Ayuda a visualizar datos univariados y bivariados.

- **Plotly.express** es otra librería para visualización con gráficos interactivos que muestran algunos detalles de las gráficas.

Vamos a cargar las librerías

In [ ]:
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sb

## <FONT SIZE=4 COLOR="green"> 7.1. TABLA DE FRECUENCIA - BARRAS Y TORTAS </FONT>

Para hacer el gráfico de barras o de tortas haremos primero una tabla de frecuencias.

- Recordemos de los gráficos de barras o de tortas se usan en general para variables categóricas o para variables cuantitativas con pocos valores, como es el caso de *número de hijos*.

- En el eje $x$ van las categorías

- En el eje $y$ la cantidad o frecuencia de cada clase o categoria

- En los diagramas de barras las barras van separadas y de colocan de diferente color.

In [ ]:
px.bar(hospital.HOSPITALIZACION.value_counts(),
       color = ["No" , "Si"])

In [ ]:
# A partir de la tabla de frecuencias de la variable hospitalización.
freq = hospital.HOSPITALIZACION.value_counts()
px.pie(freq,                               # tabla de frecuencias
       values= freq.values,                # valores de las frecuencias absolutas
       names= freq.index)                  # nombres de las categorias


## <FONT SIZE=4 COLOR="green"> 7.2 HISTOGRAMA </FONT>

- El histograma se usa para variables de tipo cuantitativo continuo o discreto con muchas mediciones diferentes (como la edad en ciertos casos).

- En el eje $x$ se identifican las clases (intervalos) de la variable.

- En el eje $y$ se identifica la frecuencia de cada una de las clases.

- Las barras en el histograma quedan contiguas y además se utiliza un solo color en estas.

In [ ]:
# Histograma más básico
px.histogram(hospital.EDAD)

Podemos colocarle más elementos de edición

In [ ]:
fig = px.histogram(hospital,                               # variable
                   nbins = 30,                             # número de intervalos
                   x = "EDAD",                             # nombre de la variable x
                   title = "Histograma con edición"        # título del gráfico
                                      )

fig.update_traces(marker_color = 'cyan',                    # color de relleno
                  marker_line_color = 'red',                # color del borde
                  marker_line_width = 1,                    # grueso de la línea
                  opacity = 1)                              # transparencia del color

fig.show()

## <FONT SIZE=4 COLOR="green"> 7.3. BOXPLOT (GRÁFICO DE CAJAS Y BIGOTES) </FONT>

- El gráfico de caja, boxplot del inglés, fue ideado por Jhon Tukey en 1977 y es una representación estadística que resalta principalmente aspectos de la distribución de los datos u observaciones por medio de las medidas de posición.

- Los diagramas de caja brindan información valiosa que no se observa en un histograma como son los cuartiles.

A continuación a describir cada uno de los elementos que tiene un boxplot

**La caja rectangular** : el gráfico contiene una caja rectangular. Es importante indicar que esta caja no tiene coordenadas cartesianas en el eje $X$  y las alturas en el eje $y$ no representa una escala de frecuencias. Las líneas horizontales en el diagrama de cajas describen el primer y tercer cuartil. $Q_1$ y $Q_3$ respectivamente.

**Mediana** : es una línea que se muestra dentro de la caja rectangular y hace referencia al valor que divide los datos en dos partes iguales. Recuerde que la mediana es igual a $Q_2$.

**Bigotes** : son líneas que salen a los costados de la caja y se usan como referencia para ubicar las observaciones que estan por fuera del $50 \%$ central de distribucion. Los bigotes nos dan información sobre la existencia de valores atípicos (*outliers*).

In [ ]:
# boxplot clásico
px.box(hospital.EDAD)

Podemos comenzar a hacer gráficos más eleborados que relacionen más variables

In [ ]:
px.box(hospital,                                                # nombre de los datos
       x = "HOSPITALIZACION",                                    # variable cualitativa o categórica
       y = "EDAD",                                              # variable cuantitativa
       title = "Boxplot variable EDAD vs HOSPITALIZACIÓN",      # título del gráfico
       color = "HOSPITALIZACION"                                # color por la variable categórica
       )

In [ ]:
px.box(hospital,                                                                # nombre de los datos
       y = "EDAD",                                                              # variable cuantitativa
       x = "ENF. CRONICA PULMONAR OBSTRUCTIVA",                                 # variable cualitativa o categórica
       title = "Boxplot variable EDAD vs ENF. CRONICA PULMONAR OBSTRUCTIVA",    # título del gráfico
       color = "ENF. CRONICA PULMONAR OBSTRUCTIVA"                              # color por la variable categórica
        )

## <FONT SIZE=4 COLOR="green"> 7.4 DIAGRAMA DE DISPERSIÓN </FONT>

- Los diagramas de dispersión permiten representar gráficamente dos variables cuantitativas.

In [ ]:
px.scatter(hospital,
           x= "EDAD",
           y= "PSA")

**Ejercicio**

Aplicar el siguiente código al conjunto apartamentos.csv.

In [ ]:
pip install sweetviz

In [ ]:
import sweetviz as sw

In [ ]:
reporte=sw.analyze(aptos, target_feat="precio")

In [ ]:
from IPython.display import HTML
reporte.show_html()
HTML("SWEETVIZ_REPORT.html")

# <FONT SIZE=5 COLOR="green"> 8. Adicional: Acceder a los elementos del DataFrame </FONT>

Para acceder a los elementos, o a las filas o columnas del DataFrame podemos usar:

- ***iloc*** : para acceder por la posición
- ***loc.*** : para acceder por el nombre

Si usamos la primera instrucción

- ***df.iloc[i,j]*** : elemento de la fila i y columna j.

- ***df.iloc[i]*** : devuelve la fila i

- ***df.iloc[lista i , lista j]*** : devuelve un data.frame con las filas en lista i y columnas en columna j.

- ***df.iloc[:, j]*** : devuelve toda columna j.

Si usamos la instrucción **.loc**

- ***df.loc[nombre fila , nombre columna]***: es lo mismo anterior pero con los nombres.

- ***df[nombre columna]*** : devuelve las columnas en name columnas

- ***df.columna*** : otra forma para acceder a una columna que en el nombre no tenga espacios en blanco.

## <FONT SIZE=4 COLOR="magenta"> 8.1 Ejemplos con **.iloc** </FONT>




In [ ]:
aptos.head()


<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc1.png?raw=true" alt="centered image" width="600" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# para seleccionar un elemento de la fila i, columna j.
aptos.iloc[1,3]

In [ ]:
# para seleccionar una fila
aptos.iloc[3]

In [ ]:
# para seleccionar una columna
aptos.iloc[ : , 1]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc2.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# filas 1 y 3 y columnas 0 y 1
aptos.iloc[[1,3], [0,1]]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc3.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# filas de la 1 a la 3 , tener en cuenta que termina un índice antes.
aptos.iloc[1:4]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc4.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# columnas 1 en adelante, ojo los dos puntos significan : end
# : solos toman todas las filas
# 1: de 1 en adelante
aptos.iloc[:, 1:]

Veamos algunos ejemplos adicionales.

In [ ]:
# las columnas 2-5
aptos.iloc[: , 2:6]

In [ ]:
# para seleccionar la segunda, tercera y primera fila
aptos.iloc[[3,4,0]]

In [ ]:
# para seleccionar la segunda y cuarta columna
aptos.iloc[: , [1,3]]

## <FONT SIZE=5 COLOR="magenta">8.2 Ejemplos con **loc** </FONT>

Veamos algunos ejemplos con **.loc**.

Recordemos que en *.iloc* se busca por posición y en *.loc* por nombre.

In [ ]:
# veamos el nombre de las columnas
aptos.columns

In [ ]:
# veamos el nombre de las filas
aptos.index

In [ ]:
# selecciono la fila 0 y 1
aptos.loc[[0,1]]

In [ ]:
aptos.loc[: , "precio"]